### PR TITLE
fix bug in hadamard_product

### PR DESCRIPTION
**Pull request by Upabjojr**
As reported in:
https://stackoverflow.com/questions/74091944/why-the-hadamard-product-result-of-two-square-matrixs-is-not-desired

NO ENTRY

### CODE REVIEW

File: sympy/matrices/expressions/hadamard.py [hunk 1]
```diff
@@ -67,6 +66,9 @@ class HadamardProduct(MatrixExpr):
 
     def __new__(cls, *args, evaluate=False, check=None):
         args = list(map(sympify, args))
+        if len(args) == 0:
+            # We currently don't have a way to support one-matrices of generic dimensions:
+            raise ValueError("HadamardProduct needs at least one argument")
```
Comment (asmeurer) [inline sympy/matrices/expressions/hadamard.py:71]: We really should generalize GenericMatrix. I think the approach for handling unknown/generic shapes needs to be done much better in the matrix expressions.

File: sympy/matrices/expressions/hadamard.py [hunk 2]
```diff
@@ -37,7 +37,6 @@ def hadamard_product(*matrices):
     if len(matrices) == 1:
         return matrices[0]
     else:
-        matrices = [i for i in matrices if not i.is_Identity]
```
Comment (asmeurer) [inline sympy/matrices/expressions/hadamard.py:40]: Weird that this line was ever here in the first place. Maybe someone incorrectly copied it from MatMul?

**Comment by asmeurer** [comment]
This looks good to me. I don't understand what happened with the CI. Why was one of the runs cancelled?

File: sympy/matrices/expressions/hadamard.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/hadamard.py:71]: That is interesting. Though I would improve the assumptions system first (is there a way to represent expressions with conditions?).

Comment (Upabjojr) [reply sympy/matrices/expressions/hadamard.py:71]: Do we have a way to create an expression with assumptions? We have the `with assuming(...): ... ` form in the assumptions module, but that is not a SymPy tree-expression.

What about something similar to Wolfram Mathematica's [Assuming](https://reference.wolfram.com/language/ref/Assuming.html)?

Comment (oscarbenjamin) [reply sympy/matrices/expressions/hadamard.py:71]: There is no way to create an expression with assumptions.

Comment (asmeurer) [reply sympy/matrices/expressions/hadamard.py:71]: `assuming` is supposed to be the equivalent of `Assuming`. However, in general, the design of the new assumptions is that assumptions are separate from expressions. 

It's really still an open question of how to most cleanly handle this in the API, e.g., should every function pass assumptions around, or should we try to store the assumptions on expressions, or store them globally, or something else.

Comment (asmeurer) [reply sympy/matrices/expressions/hadamard.py:71]: Opened https://github.com/sympy/sympy/issues/24149 for this.

**Pull request merged**